python/ray/data/tests/test_clickhouse.py

```diff
@@ -0,0 +1,189 @@
+            (
+                {
+                    "columns": ["field1"],
+                    "order_by": (["field1"], False),
+                },
+                "SELECT field1 FROM default.table_name ORDER BY field1",
+            ),
+            (
+                {
+                    "columns": ["field1", "field2"],
+                    "order_by": (["field1"], True),
+                },
+                "SELECT field1, field2 FROM default.table_name ORDER BY field1 DESC",
+            ),
+            (
+                {
+                    "columns": ["field1", "field2", "field3"],
+                    "order_by": (["field1", "field2"], False),
+                },
+                "SELECT field1, field2, field3 FROM default.table_name "
+                "ORDER BY (field1, field2)",
+            ),
+            (
+                {
+                    "columns": ["field1", "field2", "field3"],
+                    "order_by": (["field1", "field2"], True),
+                },
+                "SELECT field1, field2, field3 FROM default.table_name "
+                "ORDER BY (field1, field2) DESC",
+            ),
+            (
+                {
+                    "columns": ["field1", "field2", "field3"],
+                    "order_by": (["field1", "field2", "field3"], True),
+                },
+                "SELECT field1, field2, field3 FROM default.table_name "
+                "ORDER BY (field1, field2, field3) DESC",
+            ),
+        ],
+    )
+    def test_generate_query_full(self, datasource, query_params, expected_query):
+        datasource._columns = query_params.get("columns")
+        datasource._order_by = query_params.get("order_by")
+        generated_query = datasource._generate_query()
+        assert expected_query == generated_query
+
+    @pytest.mark.parametrize("parallelism", [1, 2, 3, 4])
+    def test_get_read_tasks(self, datasource, parallelism):
+        batch1 = pa.record_batch([pa.array([1, 2, 3, 4, 5, 6, 7, 8])], names=["field1"])
+        batch2 = pa.record_batch(
+            [pa.array([9, 10, 11, 12, 13, 14, 15, 16])], names=["field1"]
+        )
+        mock_stream = MagicMock()
+        mock_client = mock.MagicMock()
+        mock_client.query_arrow_stream.return_value.__enter__.return_value = mock_stream
+        mock_stream.__iter__.return_value = [batch1, batch2]
+        datasource.MIN_ROWS_PER_READ_TASK = 4
+        datasource._init_client = MagicMock(return_value=mock_client)
+        datasource._get_estimate = MagicMock(return_value=16)
+        mock_block_accessor = mock.MagicMock()
+        mock_block_accessor.size_bytes.return_value = 100
+        mock_block_accessor.num_rows.return_value = 4
+        mock_block_accessor.schema.return_value = batch1.schema
+        datasource._get_sample_block = MagicMock(return_value=mock_block_accessor)
+        with mock.patch(
+            "ray.data.block.BlockAccessor.for_block", return_value=mock_block_accessor
+        ):
```

Comment:
Problem with this approach is that this will mock this static method for all invocations not just 1 you intend to.

Instead let's just introduce a method producing estimates you need (size, num of rows) and mock that here

Reply:
Made this change in my latest commit.